### ISSUE TITLE
assoc? Adding a few more functional operators

### ISSUE THREAD

**Issue by Inc0n** [janet-lang/spork#108]
`assoc` and `assoc-in` are analogous to `put` and `put-in` without the side effects.

There is `first` which is sometimes more convenient than `(map (fn [x] (x 0)) xs)`. However, there isn't a `second` so we have `(map first xs)` and `(map (fn [x] (x 1)) xs)`.

There may be more of them that could be useful for FP that I haven't thought of ATM.

Would it be a good idea to add them as part of spork?

**Comment by eko234** [comment]
Hello, rn I'm indeed playing with that idea, I don't know if its the best way to do it, but by design it is not going to be performant anyway :V 
```janet
(defn assoc [ds k v & ekvs]
  (cond
   (struct? ds) (struct ;(kvs ds) k v ;ekvs)
   (and (tuple? ds) (empty? ekvs) (number? k))
   (if (= k (length ds))
     (tuple ;ds v)
     (tuple ;(slice ds 0 k) v ;(slice ds (inc k) (length ds))))
   (error "ds must be a struct or a tuple and tuples' keys must be numbers, also tuples do'nt accept extra kv's")))

(defn assoc-in [ds [k & ks] v]
  (if (empty? ks)
    (assoc ds k v)
    (assoc ds k (assoc-in (get ds k {}) ks v))))

```

AFIK it works as in clojure, it is something trivial enough so people can add it to their projects by themselves, but I think it would be nice to have it at hand with spork

**Comment by Inc0n** [comment]
It's been a while, after digging through https://github.com/janet-lang/janet/issues/1501, https://github.com/janet-lang/janet/issues/1501#issuecomment-2340989789 looks promising, in summary: 
TLDR:
```
(def assoc struct/with-proto)
```
Usage (stolen from the linked comment)
```
Janet 1.36.0-dev-c82fd106 linux/x64/gcc - '(doc)' for help
repl:1:> (def m {:a 1 :b 2})
{:a 1 :b 2}
repl:2:> (def n (struct/with-proto m :c 3))
{:c 3}
repl:3:> (struct/proto-flatten n)
{:a 1 :b 2 :c 3}
repl:4:> (def o (struct/with-proto m :a 6))
{:a 6}
repl:5:> (struct/proto-flatten o)
{:a 6 :b 2}
repl:6:> m
{:a 1 :b 2} # Original struct is unchanged
```

Credit to @CFiggers

**Comment by Inc0n** [comment]
As discussed the proposals here did not outweigh the tradeoffs it incurred
